# Changelog

## [0.3] - 2018-XX-YY

### Added
- Reads now can be demultiplexed without basecalling.
- With `--polya` turned on, poly(A) tail dwell time is written
  to `sequencing_summary.txt`. More detailed information is
  written to the dumped events by the `--dump-basecalled-events`
  switch.
- Shows detailed error messages in `poreplex.log`.
- The files generated by `--dump-basecalled-events` now includes
  the attributes for signal scaling parameters and the last position
  that DNA adapter ends within the signal.
- Extremely short sequences are now sorted into `fail`ed reads.
  The minimum required length can be changed with the
  `--minimum-length` switch. (Suggested by Nathan Roach)

### Changed
- Fixed an error that stops overall process by an invalid FAST5 file.
- Fixed a problem that separate lines of FASTA, FASTQ or
  `sequencing-summary.txt` are mixed up in the output file sometimes.
- Updated the neural network model for barcode demultiplexing for
  even less false positives using randomly stitched signal fragments
  as a background.

